zerver/tests/test_typing.py

```diff
@@ -290,3 +344,75 @@ def test_stop_to_another_user(self) -> None:
         self.assertEqual(event["sender"]["email"], sender.email)
         self.assertEqual(event["type"], "typing")
         self.assertEqual(event["op"], "stop")
+
+
+class TypingHappyPathTestStreams(ZulipTestCase):
+    def test_start(self) -> None:
+        sender = self.example_user("hamlet")
+        stream_name = self.get_streams(sender)[0]
+        stream_id = self.get_stream_id(stream_name)
+        topic = "Some topic"
+
+        expected_user_ids = {
+            user_profile.id
+            for user_profile in self.users_subscribed_to_stream(stream_name, sender.realm)
+        }
+
+        params = dict(
+            type="stream",
+            op="start",
+            to=orjson.dumps([stream_id]).decode(),
+            topic=topic,
+        )
+
+        events: List[Mapping[str, Any]] = []
+        with tornado_redirected_to_list(events):
+            result = self.api_post(sender, "/api/v1/typing", params)
+
+        self.assert_json_success(result)
+        self.assertEqual(len(events), 1)
```

Comment:
Can you do query count assertions in these new tests as well?  They are somewhat important given that this is a high-volume endpoint.

Reply:
Sure, added those. It isn't clear to me how those help so I skipped them at the time of writing those thinking I'd ask about them later but forgot. Could you provide more data/ pointers to understand how query count assertions in tests help? 

To make sure what I had done is right: I just checked the length of `queries` with a `print` and added an assert to that value. Is it what is to be done, or there's something else for me to do to make sure 12 queries are run?

Reply:
Yeah, that's the right way to do it.  It's worth printing the queries once nicely one-per-line and pasting the result here, so we can confirm it's not broken.

Reply:
My previous code was inefficient in fetching users IDs. It was running a query of each user, just to get user ID. Changed it now to use an existing function `get_user_ids_for_streams`. 

These are the queries: (Not sure why there's a client = Mobile query, I presume it's somehow related to `self.api_post` of tests)

0 SELECT "zerver_userprofile"."id", "zerver_userprofile"."password", "zerver_userprofile"."last_login", "zerver_userprofile"."is_superuser", "zerver_userprofile"."delivery_email", "zerver_userprofile"."email", "zerver_userprofile"."realm_id", "zerver_userprofile"."recipient_id", "zerver_userprofile"."full_name", "zerver_userprofile"."date_joined", "zerver_userprofile"."tos_version", "zerver_userprofile"."api_key", "zerver_userprofile"."is_staff", "zerver_userprofile"."is_active", "zerver_userprofile"."is_billing_admin", "zerver_userprofile"."is_bot", "zerver_userprofile"."bot_type", "zerver_userprofile"."bot_owner_id", "zerver_userprofile"."role", "zerver_userprofile"."long_term_idle", "zerver_userprofile"."last_active_message_id", "zerver_userprofile"."is_mirror_dummy", "zerver_userprofile"."can_forge_sender", "zerver_userprofile"."can_create_users", "zerver_userprofile"."enable_stream_desktop_notifications", "zerver_userprofile"."enable_stream_email_notifications", "zerver_userprofile"."enable_stream_push_notifications", "zerver_userprofile"."enable_stream_audible_notifications", "zerver_userprofile"."notification_sound", "zerver_userprofile"."wildcard_mentions_notify", "zerver_userprofile"."enable_desktop_notifications", "zerver_userprofile"."pm_content_in_desktop_notifications", "zerver_userprofile"."enable_sounds", "zerver_userprofile"."enable_offline_email_notifications", "zerver_userprofile"."message_content_in_email_notifications", "zerver_userprofile"."enable_offline_push_notifications", "zerver_userprofile"."enable_online_push_notifications", "zerver_userprofile"."desktop_icon_count_display", "zerver_userprofile"."enable_digest_emails", "zerver_userprofile"."enable_login_emails", "zerver_userprofile"."realm_name_in_notifications", "zerver_userprofile"."presence_enabled", "zerver_userprofile"."last_reminder", "zerver_userprofile"."rate_limits", "zerver_userprofile"."default_sending_stream_id", "zerver_userprofile"."default_events_register_stream_id", "zerver_userprofile"."default_all_public_streams", "zerver_userprofile"."enter_sends", "zerver_userprofile"."left_side_userlist", "zerver_userprofile"."default_language", "zerver_userprofile"."default_view", "zerver_userprofile"."dense_mode", "zerver_userprofile"."fluid_layout_width", "zerver_userprofile"."high_contrast_mode", "zerver_userprofile"."translate_emoticons", "zerver_userprofile"."twenty_four_hour_time", "zerver_userprofile"."starred_message_counts", "zerver_userprofile"."color_scheme", "zerver_userprofile"."demote_inactive_streams", "zerver_userprofile"."timezone", "zerver_userprofile"."emojiset", "zerver_userprofile"."avatar_source", "zerver_userprofile"."avatar_version", "zerver_userprofile"."avatar_hash", "zerver_userprofile"."tutorial_status", "zerver_userprofile"."onboarding_steps", "zerver_userprofile"."zoom_token", "zerver_realm"."id", "zerver_realm"."name", "zerver_realm"."description", "zerver_realm"."string_id", "zerver_realm"."date_created", "zerver_realm"."deactivated", "zerver_realm"."deactivated_redirect", "zerver_realm"."emails_restricted_to_domains", "zerver_realm"."invite_required", "zerver_realm"."max_invites", "zerver_realm"."disallow_disposable_email_addresses", "zerver_realm"."authentication_methods", "zerver_realm"."inline_image_preview", "zerver_realm"."inline_url_embed_preview", "zerver_realm"."digest_emails_enabled", "zerver_realm"."digest_weekday", "zerver_realm"."send_welcome_emails", "zerver_realm"."message_content_allowed_in_email_notifications", "zerver_realm"."mandatory_topics", "zerver_realm"."add_emoji_by_admins_only", "zerver_realm"."name_changes_disabled", "zerver_realm"."email_changes_disabled", "zerver_realm"."avatar_changes_disabled", "zerver_realm"."create_stream_policy", "zerver_realm"."invite_to_realm_policy", "zerver_realm"."invite_to_stream_policy", "zerver_realm"."move_messages_between_streams_policy", "zerver_realm"."user_group_edit_policy", "zerver_realm"."private_message_policy", "zerver_realm"."wildcard_mention_policy", "zerver_realm"."email_address_visibility", "zerver_realm"."waiting_period_threshold", "zerver_realm"."allow_message_deleting", "zerver_realm"."message_content_delete_limit_seconds", "zerver_realm"."allow_message_editing", "zerver_realm"."message_content_edit_limit_seconds", "zerver_realm"."allow_edit_history", "zerver_realm"."allow_community_topic_editing", "zerver_realm"."default_twenty_four_hour_time", "zerver_realm"."default_language", "zerver_realm"."notifications_stream_id", "zerver_realm"."signup_notifications_stream_id", "zerver_realm"."message_retention_days", "zerver_realm"."message_visibility_limit", "zerver_realm"."first_visible_message_id", "zerver_realm"."org_type", "zerver_realm"."plan_type", "zerver_realm"."bot_creation_policy", "zerver_realm"."upload_quota_gb", "zerver_realm"."video_chat_provider", "zerver_realm"."giphy_rating", "zerver_realm"."default_code_block_language", "zerver_realm"."icon_source", "zerver_realm"."icon_version", "zerver_realm"."logo_source", "zerver_realm"."logo_version", "zerver_realm"."night_logo_source", "zerver_realm"."night_logo_version" FROM "zerver_userprofile" INNER JOIN "zerver_realm" ON ("zerver_userprofile"."realm_id" = "zerver_realm"."id") WHERE "zerver_userprofile"."api_key" = 'B7mRHbe3e8yHJOOuBKk6cf5fWnO2Bsj6' LIMIT 21

1 SELECT "zerver_client"."id", "zerver_client"."name" FROM "zerver_client" WHERE "zerver_client"."name" = 'ZulipMobile' LIMIT 21

2 SELECT "zerver_stream"."id", "zerver_stream"."name", "zerver_stream"."realm_id", "zerver_stream"."date_created", "zerver_stream"."deactivated", "zerver_stream"."description", "zerver_stream"."rendered_description", "zerver_stream"."recipient_id", "zerver_stream"."invite_only", "zerver_stream"."history_public_to_subscribers", "zerver_stream"."is_web_public", "zerver_stream"."stream_post_policy", "zerver_stream"."is_in_zephyr_realm", "zerver_stream"."email_token", "zerver_stream"."message_retention_days", "zerver_stream"."first_message_id", "zerver_realm"."id", "zerver_realm"."name", "zerver_realm"."description", "zerver_realm"."string_id", "zerver_realm"."date_created", "zerver_realm"."deactivated", "zerver_realm"."deactivated_redirect", "zerver_realm"."emails_restricted_to_domains", "zerver_realm"."invite_required", "zerver_realm"."max_invites", "zerver_realm"."disallow_disposable_email_addresses", "zerver_realm"."authentication_methods", "zerver_realm"."inline_image_preview", "zerver_realm"."inline_url_embed_preview", "zerver_realm"."digest_emails_enabled", "zerver_realm"."digest_weekday", "zerver_realm"."send_welcome_emails", "zerver_realm"."message_content_allowed_in_email_notifications", "zerver_realm"."mandatory_topics", "zerver_realm"."add_emoji_by_admins_only", "zerver_realm"."name_changes_disabled", "zerver_realm"."email_changes_disabled", "zerver_realm"."avatar_changes_disabled", "zerver_realm"."create_stream_policy", "zerver_realm"."invite_to_realm_policy", "zerver_realm"."invite_to_stream_policy", "zerver_realm"."move_messages_between_streams_policy", "zerver_realm"."user_group_edit_policy", "zerver_realm"."private_message_policy", "zerver_realm"."wildcard_mention_policy", "zerver_realm"."email_address_visibility", "zerver_realm"."waiting_period_threshold", "zerver_realm"."allow_message_deleting", "zerver_realm"."message_content_delete_limit_seconds", "zerver_realm"."allow_message_editing", "zerver_realm"."message_content_edit_limit_seconds", "zerver_realm"."allow_edit_history", "zerver_realm"."allow_community_topic_editing", "zerver_realm"."default_twenty_four_hour_time", "zerver_realm"."default_language", "zerver_realm"."notifications_stream_id", "zerver_realm"."signup_notifications_stream_id", "zerver_realm"."message_retention_days", "zerver_realm"."message_visibility_limit", "zerver_realm"."first_visible_message_id", "zerver_realm"."org_type", "zerver_realm"."plan_type", "zerver_realm"."bot_creation_policy", "zerver_realm"."upload_quota_gb", "zerver_realm"."video_chat_provider", "zerver_realm"."giphy_rating", "zerver_realm"."default_code_block_language", "zerver_realm"."icon_source", "zerver_realm"."icon_version", "zerver_realm"."logo_source", "zerver_realm"."logo_version", "zerver_realm"."night_logo_source", "zerver_realm"."night_logo_version" FROM "zerver_stream" INNER JOIN "zerver_realm" ON ("zerver_stream"."realm_id" = "zerver_realm"."id") WHERE ("zerver_stream"."id" = 1 AND "zerver_stream"."realm_id" = 2) LIMIT 21

3 SELECT "zerver_subscription"."id", "zerver_subscription"."user_profile_id", "zerver_subscription"."recipient_id", "zerver_subscription"."active", "zerver_subscription"."is_user_active", "zerver_subscription"."role", "zerver_subscription"."is_muted", "zerver_subscription"."color", "zerver_subscription"."pin_to_top", "zerver_subscription"."desktop_notifications", "zerver_subscription"."audible_notifications", "zerver_subscription"."push_notifications", "zerver_subscription"."email_notifications", "zerver_subscription"."wildcard_mentions_notify" FROM "zerver_subscription" WHERE ("zerver_subscription"."active" AND "zerver_subscription"."recipient_id" = 120 AND "zerver_subscription"."user_profile_id" = 10) LIMIT 21

4 SELECT "zerver_recipient"."type_id", "zerver_subscription"."user_profile_id" FROM "zerver_subscription" INNER JOIN "zerver_recipient" ON ("zerver_subscription"."recipient_id" = "zerver_recipient"."id") INNER JOIN "zerver_userprofile" ON ("zerver_subscription"."user_profile_id" = "zerver_userprofile"."id") WHERE ("zerver_subscription"."active" AND "zerver_recipient"."type" = 2 AND "zerver_recipient"."type_id" IN (1) AND "zerver_userprofile"."is_active") ORDER BY "zerver_recipient"."type_id" ASC